src/profile-logic/symbol-store.js

```diff
@@ -5,149 +5,296 @@
-            this._failedRequests.add(libid);
-            this._requestedSymbolTables.delete(libid);
-            throw new SymbolsNotFoundError(
-              `Failed to symbolicate library ${debugName}`,
-              lib,
-              error
-            );
-          });
-
-        // Once the symbol table comes in, store it in the database, but don't
-        // let that block the promise that we return to our caller.
-        symbolTablePromise.then(symbolTable => {
-          this._db.storeSymbolTable(debugName, breakpadId, symbolTable).then(
-            () => {
-              this._requestedSymbolTables.delete(libid);
-            },
-            error => {
-              console.error(
-                `Failed to store the symbol table for ${debugName} ${breakpadId} in the database:`,
-                error
-              );
-              // We'll keep the symbolTablePromise in _requestedSymbolTables so
-              // that we'll the symbolTable around for future requests even though
-              // we failed to put it into the database.
-            }
-          );
+        results.set(address, {
+          functionOffset: address - symbolTableAddrs[symbolIndex],
+          name: currentSymbol,
         });
-
-        return symbolTablePromise;
-      });
-    this._requestedSymbolTables.set(libid, symbolTablePromise);
-    return symbolTablePromise;
+      } else {
+        results.set(address, {
+          functionOffset: address,
+          name: '<before first symbol>',
+        });
+      }
+    }
+    return results;
   }
 
   /**
-   * Get the array of symbol addresses for the given library.
-   * @param  {Library} lib A library object with the properties `debugName` and `breakpadId`.
-   * @return {Promise<number[]>} A promise of the array of addresses.
+   * Look up symbols for the given addresses in the given libraries.
+   * For each LibSymbolicationRequest in requests, either errorCb or successCb
+   * will be called with that LibSymbolicationRequest and the result / error.
+   * This method returns a promise that resolves when the callbacks for all
+   * requests have been called.
    */
-  async getFuncAddressTableForLib(lib: Library): Promise<Uint32Array> {
-    const [addrs] = await this._getSymbolTable(lib);
-    return addrs;
-  }
+  async getSymbols(
+    requests: LibSymbolicationRequest[],
+    successCb: (LibSymbolicationRequest, Map<number, AddressResult>) => void,
+    errorCb: (LibSymbolicationRequest, Error) => void
+  ): Promise<void> {
+    const requestsForNonCachedLibs = [];
+    const requestsForCachedLibs = [];
+    await Promise.all(
+      requests.map(async request => {
+        // Try to get the symbol table from the database.
+        const { debugName, breakpadId } = request.lib;
+        try {
+          requestsForCachedLibs.push({
+            request,
+            symbolTable: await this._db.getSymbolTable(debugName, breakpadId),
+          });
```

Comment:
nit: I find this difficult to read: if we don't find the data in the db, this `await` will throw and nothing will be actually added in `requestsForCachedLibs`. But reading this code this is not obvious: I had to think really hard to understand this.

How about decomposing this in 2 steps, like this:
```javascript
const symbolTable = await this._db.getSymbolTable(debugName, breakpadId);
requestsForCachedLibs.push({ request, symbolTable });
```